From 03694fa6a2f0f71d9737f95a7a98851842763254 Mon Sep 17 00:00:00 2001
From: Bjorn Pagen <hello@bjornpagen.com>
Date: Sun, 9 Aug 2026 10:54:17 -0500
Subject: [PATCH] fixincludes: Fix rsize_t with Darwin modules [PR999999]

The Darwin SDK selects a Clang-specific branch when
__has_feature(modules) is true.  GCC also reports this feature with
-fmodules.  GCC stddef.h does not support the __need_rsize_t protocol.
The SDK therefore leaves rsize_t undefined.  SDK declarations then
fail to parse.

Make the SDK guard require __clang__.  GCC then uses the SDK typedef
branch.  This change does not change Clang behavior.

fixincludes/ChangeLog:

	PR target/999999
	* inclhack.def (darwin_rsize_t_modules): New fix.
	* fixincl.x: Regenerate.
	* tests/base/sys/_types/_rsize_t.h: New test.

Signed-off-by: Bjorn Pagen <hello@bjornpagen.com>
---
 fixincludes/fixincl.x                        | 60 ++++++++++++++++++--
 fixincludes/inclhack.def                     | 26 +++++++++
 fixincludes/tests/base/sys/_types/_rsize_t.h | 18 ++++++
 3 files changed, 99 insertions(+), 5 deletions(-)
 create mode 100644 fixincludes/tests/base/sys/_types/_rsize_t.h

diff --git a/fixincludes/fixincl.x b/fixincludes/fixincl.x
index 9db4d376a2e..99733534a9c 100644
--- a/fixincludes/fixincl.x
+++ b/fixincludes/fixincl.x
@@ -2,11 +2,11 @@
  *
  * DO NOT EDIT THIS FILE   (fixincl.x)
  *
- * It has been AutoGen-ed  February 15, 2026 at 11:11:08 PM by AutoGen 5.18.12
+ * It has been AutoGen-ed  August 10, 2026 at 09:49:15 AM by AutoGen 5.18.16
  * From the definitions    inclhack.def
  * and the template file   fixincl
  */
-/* DO NOT MERGE THIS FILE, EITHER Sun Feb 15 23:11:08 CET 2026
+/* DO NOT MERGE THIS FILE, EITHER Mon Aug 10 09:49:15 UTC 2026
  *
  * You must regenerate it.  Use the ./genfixes script.
  *
@@ -15,7 +15,7 @@
  * certain ANSI-incompatible system header files which are fixed to work
  * correctly with ANSI C and placed in a directory that GNU C will search.
  *
- * This file contains 251 fixup descriptions.
+ * This file contains 252 fixup descriptions.
  *
  * See README for more information.
  *
@@ -3287,6 +3287,50 @@ static const char* apzDarwin_Private_ExternPatch[] = {
     "__private_extern__",
     (char*)NULL };
 
+/* * * * * * * * * * * * * * * * * * * * * * * * * *
+ *
+ *  Description of Darwin_Rsize_T_Modules fix
+ */
+tSCC zDarwin_Rsize_T_ModulesName[] =
+     "darwin_rsize_t_modules";
+
+/*
+ *  File name selection pattern
+ */
+tSCC zDarwin_Rsize_T_ModulesList[] =
+  "sys/_types/_rsize_t.h\0";
+/*
+ *  Machine/OS name selection pattern
+ */
+tSCC* apzDarwin_Rsize_T_ModulesMachs[] = {
+        "*-*-darwin*",
+        (const char*)NULL };
+
+/*
+ *  content selection pattern - do fix if pattern found
+ */
+tSCC zDarwin_Rsize_T_ModulesSelect0[] =
+       "defined\\(__has_feature\\) && __has_feature\\(modules\\)";
+
+/*
+ *  content bypass pattern - skip fix if pattern found
+ */
+tSCC zDarwin_Rsize_T_ModulesBypass0[] =
+       "defined\\(__clang__\\) && defined\\(__has_feature\\) && __has_feature\\(modules\\)";
+
+#define    DARWIN_RSIZE_T_MODULES_TEST_CT  2
+static tTestDesc aDarwin_Rsize_T_ModulesTests[] = {
+  { TT_NEGREP,   zDarwin_Rsize_T_ModulesBypass0, (regex_t*)NULL },
+  { TT_EGREP,    zDarwin_Rsize_T_ModulesSelect0, (regex_t*)NULL }, };
+
+/*
+ *  Fix Command Arguments for Darwin_Rsize_T_Modules
+ */
+static const char* apzDarwin_Rsize_T_ModulesPatch[] = {
+    "format",
+    "defined(__clang__) && defined(__has_feature) && __has_feature(modules)",
+    (char*)NULL };
+
 /* * * * * * * * * * * * * * * * * * * * * * * * * *
  *
  *  Description of Darwin_Stdint_1 fix
@@ -10283,9 +10327,9 @@ static const char* apzX11_SprintfPatch[] = {
  *
  *  List of all fixes
  */
-#define REGEX_COUNT          291
+#define REGEX_COUNT          293
 #define MACH_LIST_SIZE_LIMIT 187
-#define FIX_COUNT            251
+#define FIX_COUNT            252
 
 /*
  *  Enumerate the fixes
@@ -10368,6 +10412,7 @@ typedef enum {
     DARWIN_OS_BASE_1_FIXIDX,
     DARWIN_DISPATCH_OBJECT_1_FIXIDX,
     DARWIN_PRIVATE_EXTERN_FIXIDX,
+    DARWIN_RSIZE_T_MODULES_FIXIDX,
     DARWIN_STDINT_1_FIXIDX,
     DARWIN_STDINT_2_FIXIDX,
     DARWIN_STDINT_3_FIXIDX,
@@ -10930,6 +10975,11 @@ tFixDesc fixDescList[ FIX_COUNT ] = {
      DARWIN_PRIVATE_EXTERN_TEST_CT, FD_MACH_ONLY | FD_SUBROUTINE,
      aDarwin_Private_ExternTests,   apzDarwin_Private_ExternPatch, 0 },
 
+  {  zDarwin_Rsize_T_ModulesName,    zDarwin_Rsize_T_ModulesList,
+     apzDarwin_Rsize_T_ModulesMachs,
+     DARWIN_RSIZE_T_MODULES_TEST_CT, FD_MACH_ONLY | FD_SUBROUTINE,
+     aDarwin_Rsize_T_ModulesTests,   apzDarwin_Rsize_T_ModulesPatch, 0 },
+
   {  zDarwin_Stdint_1Name,    zDarwin_Stdint_1List,
      apzDarwin_Stdint_1Machs,
      DARWIN_STDINT_1_TEST_CT, FD_MACH_ONLY | FD_SUBROUTINE,
diff --git a/fixincludes/inclhack.def b/fixincludes/inclhack.def
index 20a7e8ef977..f8e710c4697 100644
--- a/fixincludes/inclhack.def
+++ b/fixincludes/inclhack.def
@@ -1602,6 +1602,32 @@ fix = {
 	      "unsigned long *address);\n";
 };
 
+/*
+ * The SDK's <sys/_types/_rsize_t.h> gates its clang-modules branch on
+ * __has_feature(modules), but GCC also reports that feature under
+ * -fmodules.  The branch then relies on the clang-only stddef.h
+ * __need_rsize_t protocol, which GCC's stddef.h does not implement, so
+ * rsize_t is never defined and the C11 Annex K declarations in the SDK's
+ * <string.h> fail to parse.  Require __clang__ so that GCC always takes
+ * the plain typedef branch.
+ */
+fix = {
+    hackname  = darwin_rsize_t_modules;
+    mach      = "*-*-darwin*";
+    files     = sys/_types/_rsize_t.h;
+    bypass    = "defined\\(__clang__\\) && defined\\(__has_feature\\) && "
+		"__has_feature\\(modules\\)";
+    select    = "defined\\(__has_feature\\) && __has_feature\\(modules\\)";
+    c_fix     = format;
+    c_fix_arg = "defined(__clang__) && defined(__has_feature) && "
+		"__has_feature(modules)";
+    test_text = "#if defined(__has_feature) && __has_feature(modules)\n"
+		"#define USE_CLANG_STDDEF 1\n"
+		"#else\n"
+		"#define USE_CLANG_STDDEF 0\n"
+		"#endif";
+};
+
 /*
  * Darwin headers have a stdint.h that defines UINT8_C and UINT16_C to
  * unsigned constants.
diff --git a/fixincludes/tests/base/sys/_types/_rsize_t.h b/fixincludes/tests/base/sys/_types/_rsize_t.h
new file mode 100644
index 00000000000..7ff637d9d1a
--- /dev/null
+++ b/fixincludes/tests/base/sys/_types/_rsize_t.h
@@ -0,0 +1,18 @@
+/*  DO NOT EDIT THIS FILE.
+
+    It has been auto-edited by fixincludes from:
+
+	"fixinc/tests/inc/sys/_types/_rsize_t.h"
+
+    This had to be done to correct non-standard usages in the
+    original, manufacturer supplied header file.  */
+
+
+
+#if defined( DARWIN_RSIZE_T_MODULES_CHECK )
+#if defined(__clang__) && defined(__has_feature) && __has_feature(modules)
+#define USE_CLANG_STDDEF 1
+#else
+#define USE_CLANG_STDDEF 0
+#endif
+#endif  /* DARWIN_RSIZE_T_MODULES_CHECK */
-- 
2.50.1 (Apple Git-155)

